Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/non_rigid_mesh_registration.h

```diff
@@ -0,0 +1,930 @@
+  using Neighbor_search = CGAL::Orthogonal_k_neighbor_search<Search_traits>;
+  using KDTree = typename Neighbor_search::Tree;
+
+  Eigen_matrix_to_point_map a(points);
+
+  KDTree kdtree(boost::counting_iterator<std::size_t>(0), boost::counting_iterator<std::size_t>(points.rows()), KDTree::Splitter(), Search_traits(Eigen_matrix_to_point_map(points)));
+  kdtree.build();
+
+  Eigen::MatrixXi idz(query.rows(), k);
+  Eigen::MatrixXf dist(query.rows(), k);
+
+  for (Index i = 0; i < query.rows(); ++i) {
+    Point_3 query_pt = { query(i, 0), query(i, 1), query(i, 2) };
+    Neighbor_search search(kdtree, query_pt, k, 0, true, Neighbor_search::Distance(Eigen_matrix_to_point_map(points)));
+    Index j = 0;
+    for (auto it = search.begin(); it != search.end() && j < k; ++it) {
+      idz(i, j) = it->first;
+      dist(i, j) = it->second;
+    }
+  }
+
+  return std::make_pair(idz, dist);
+}
+
+template <typename T>
+int sign(T val) {
+  return (T(0) < val) - (val < T(0));
+}
+
+void insertSparseMatrix(const SparseMat& mat, std::vector<SparseTriplet>& coefficients, size_t start_i = 0, size_t start_j = 0) {
+  for (int k = 0; k < mat.outerSize(); ++k)
+    for (SparseMat::InnerIterator it(mat, k); it; ++it)
+      coefficients.push_back(SparseTriplet(start_i + it.row(), start_j + it.col(), it.value()));
+}
+
+template<typename T, typename Visitor, typename TriangleMesh>
+void rotation(std::size_t index, Visitor &v, TriangleMesh &source, const Vertices& X, const Vertices& Z, const std::set<int>& neighbors, const std::vector<T> &weights, std::vector<Eigen::Matrix<ScalarType, 3, 3>> &rotations) {
+  using Vertex_index = typename boost::graph_traits<TriangleMesh>::vertex_descriptor;
+  Deformation_Eigen_closest_rotation_traits_3 cr;
+
+  Eigen::Matrix3d cov;
+  cov.setZero();
+
+  Eigen::Vector3d x_src = X.row(index), z_src = Z.row(index);
+  auto n = neighbors.begin();
+  auto w = weights.begin();
+
+  v.rotation_matrix_pre(Vertex_index(index), source);
+
+  for (; (n != neighbors.end()) && (w != weights.end()); n++, w++) {
+    Eigen::Vector3d x_dst = X.row(*n);
+    Eigen::Vector3d z_dst = Z.row(*n);
+    Eigen::Vector3d source_edge = x_src - x_dst;
+    Eigen::Vector3d moving_edge = z_src - z_dst;
+
+    cr.add_scalar_t_vector_t_vector_transpose(cov, *w, source_edge, moving_edge);
+    v.update_covariance_matrix(cov, rotations[index]);
+  }
+
+  cr.compute_close_rotation(cov, rotations[index]);
+}
+
+template <typename T>
+Eigen::Matrix<T, 3, 3> rot(T a, T b, T c) {
+  T ca = cos(a), cb = cos(b), cc = cos(c);
+  T sa = sin(a), sb = sin(b), sc = sin(c);
+  Eigen::Matrix<T, 3, 3> R;
+  R << cb * cc, cc* sa* sb - ca * sc, ca* cc* sb + sa * sc,
+    cb* sc, ca* cc + sa * sb * sc, ca* sb* sc - cc * sa,
+    -sb, cb* sa, ca* cb;
+
+  return R;
+}
+
+#endif
+} // namespace registration
+} // namespace internal
+
+/*!
+* \ingroup PMP_registration_grp
+*
+* \brief computes non-rigid transformation of a mesh onto a set of oriented points.
```

Comment:
is it clear that it means points + oriented normals?

Reply:
I've read it before. But in the CGAL repository the only other occurrence is in the poisson reconstruction documentation.